### PR TITLE
Remove concept of window backend

### PR DESCRIPTION
**Pull request by peppy**
- Closes https://github.com/ppy/osu-framework/issues/3990
- [x] Depends on #3995
- [x] Need to figure why external resizes aren't updating the reported size immediately
- [x] Properly handle fullscreen support (doesn't work on master anyway soo)

This fixes too many issues to list, and makes too many refactors to explain. Not sure where to start, but recommend reviewing this using full file passes, not diff. Have tested that mostly everything works (equal to or more than previously) apart from the listed task list above.

This is by no means a final pass on this code. Just trying to get things into a roughly manageable state to figure out the remaining issues and continue to improve the readability.

### CODE REVIEW

**Review by bdach — COMMENTED** [review]
Skimmed through this once. Seems saner although I will definitely have missed things as `DesktopWindow` is still pretty bad reading due to everything being in one class, sometimes intertwined between responsibilities. Extracting partials might be a future step for this code.

The code runs on linux, but I suspect it might not be working properly with the osuTK raw mouse handler in windowed mode (seems to offset mouse position by position of window top-left corner).

File: osu.Framework/Platform/DesktopWindow.cs [hunk 1]
```diff
@@ -2,107 +2,1054 @@
...
 namespace osu.Framework.Platform
 {
     /// <summary>
-    /// Implementation of <see cref="Window"/> used for desktop platforms.
-    /// Uses <see cref="Sdl2WindowBackend"/> and <see cref="Sdl2GraphicsBackend"/> by default.
+    /// Implementation of <see cref="IWindow"/> that provides bindables and
+    /// delegates responsibility to window and graphics backends.
     /// </summary>
-    public class DesktopWindow : Window
+    public class DesktopWindow : IWindow
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Platform/DesktopWindow.cs:31]: I'm open to changing it to this. Really depends on whether we consider sdl the "default" or just one of many options.

File: SampleGame.Desktop/Program.cs [hunk 2]
```diff
@@ -13,9 +13,9 @@ public static class Program
         [STAThread]
         public static void Main(string[] args)
         {
-            bool useSdl = args.Contains(@"--sdl");
+            bool useOpenTK = args.Contains(@"--tk");
 
-            using (GameHost host = Host.GetSuitableHost(@"sample-game", useSdl: useSdl))
+            using (GameHost host = Host.GetSuitableHost(@"sample-game", useSdl: !useOpenTK))
```
[the remark this answers is not in the document]
Comment (peppy) [reply SampleGame.Desktop/Program.cs:18]: hmm, could do. not sure.

File: osu.Framework/Platform/DesktopWindow.cs [hunk 3]
```diff
@@ -2,107 +2,1054 @@
...
+        /// Gets the native window handle as provided by the operating system.
+        /// </summary>
+        public IntPtr WindowHandle
+        {
+            get
+            {
+                if (SdlWindowHandle == IntPtr.Zero)
+                    return IntPtr.Zero;
+
+                var wmInfo = getWindowWMInfo;
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Platform/DesktopWindow.cs:288]: oops, was supposed to be a method.

**Comment by peppy** [comment]
I've fixed up all the remaining issues I know of. Fullscreen now works correctly on windows (borderless not so much, yet). It doesn't work on macOS but that can be fixed later, needs further investigation.

File: osu.Framework/Input/Handlers/Keyboard/KeyboardHandler.cs [hunk 4]
```diff
@@ -16,7 +16,7 @@ public class KeyboardHandler : InputHandler
 
         public override bool Initialize(GameHost host)
         {
-            if (!(host.Window is DesktopWindow window))
+            if (!(host.Window is SDL2DesktopWindow window))
```
Comment (bdach) [inline osu.Framework/Input/Handlers/Keyboard/KeyboardHandler.cs:19]: This is something I didn't anticipate with the rename of `DesktopWindow` - now every seemingly generic keyboard/joystick handler refuses to initialise with anything but SDL. I wouldn't muddle this diff with changes in that regard but a next step would be probably to rename these handlers to match as they're windowing-toolkit-specific as they are anyway.

File: osu.Framework/Host.cs [hunk 5]
```diff
@@ -12,7 +12,7 @@ namespace osu.Framework
 {
     public static class Host
     {
-        public static DesktopGameHost GetSuitableHost(string gameName, bool bindIPC = false, bool portableInstallation = false, bool useSdl = false)
+        public static DesktopGameHost GetSuitableHost(string gameName, bool bindIPC = false, bool portableInstallation = false, bool useOsuTK = true)
```
Comment (bdach) [inline osu.Framework/Host.cs:15]: Hmm... so all existing framework consumers stay on osuTK, but get `(legacy osuTK)` tacked onto their window names... Not saying that's bad, just a thought that occurred in my head.

**Review by bdach — COMMENTED** [review]
Looks mostly good to go, just a couple of notes in passing.

File: osu.Framework/Host.cs (hunk 5 above)
Comment (peppy) [reply osu.Framework/Host.cs:15]: oh this should be false, good catch.

File: osu.Framework/Input/Handlers/Keyboard/KeyboardHandler.cs (hunk 4 above)
Comment (peppy) [reply osu.Framework/Input/Handlers/Keyboard/KeyboardHandler.cs:19]: another good catch.

Comment (peppy) [reply osu.Framework/Input/Handlers/Keyboard/KeyboardHandler.cs:19]: Touching on this again now that i've checked the code: I think the path forward here is replacing the conditions with `IWindow`, or removing it altogether.

The only reason this cannot be done is due to the existence of `OsuTK` legacy classes, which aren't hooked up correctly with the new event flow. Once that is removed all new `IWindow` implementations should be able to use the default input handlers in most cases.

Comment (bdach) [reply osu.Framework/Input/Handlers/Keyboard/KeyboardHandler.cs:19]: Yep, definitely could definitely agree with either. Any way you cut it this is not a thing to be fixed by this PR.

**Review by smoogipoo — APPROVED** [review]
Let's do this

**Pull request merged**